src/Modules/OrchardCore.Commerce/Services/LocalInventoryProvider.cs

```diff
@@ -0,0 +1,73 @@
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.Inventory.Models;
+using OrchardCore.Commerce.Models;
+using OrchardCore.ContentManagement;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using YesSql;
+
+namespace OrchardCore.Commerce.Services;
+
+/// <summary>
+/// A product inventory provider that updates inventories of shopping cart items which have an <c>InventoryPart</c>.
+/// </summary>
+public class LocalInventoryProvider : IProductInventoryProvider
+{
+    private readonly IProductService _productService;
+    private readonly ISession _session;
+    private readonly SemaphoreSlim _lock = new(initialCount: 1);
```

Comment:
I think the SemaphoreSlim is supposed to be static, since it's supposed to lock across threads.